### ISSUE TITLE
Getting regression coefficients for additional regressors

### ISSUE THREAD

**Issue by AnupaS** [facebook/prophet#848]
Hi,
How can I get the regression coefficients for additional regressors added in prophet. I am using the below code, using two regressors. If I look at beta values in params, it has 28 values. Is it possible to get regression coefficients from them?

```
    m = Prophet(changepoint_prior_scale=cp, 
            weekly_seasonality=False, daily_seasonality=False)
    m.add_seasonality(name='monthly', period=30, fourier_order=3)
    m.add_regressor('Price per Unit')
    m.add_regressor('Promo_present')

```

**Comment by chessplayer2** [comment]
Hi Anupa,

You will first need to forecast for your regressors. For example, you can create a prophet model for each one of your regressors to achieve this.

Once you have a forecast for your regressors, you will need to put them in a dataframe which will then be used to predict your outcome variable. Once you do that, you will be able to create a dataframe with the regression coefficients using the predict function.

Example:

data_regressors <- data.frame(
  ds               = mydata$ds, --dates of actual and forecasted data

  Price_per_Unit      = Forecast_Price_per_Unit$yhat, --this will be coming from the model you use to forecast 'Price per Unit'

  Promo_present     = Promo_present$yhat) --this will be coming from the model you use to forecast 'Promo_present'

--Run Model
    m = Prophet(changepoint_prior_scale=cp, 
            weekly_seasonality=False, daily_seasonality=False)
    m.add_seasonality(name='monthly', period=30, fourier_order=3)
    m.add_regressor('Price_per_Unit')
    m.add_regressor('Promo_present')


forecasted_data <- predict(m, data_regressors)
head(forecasted_data)

**Comment by AnupaS** [comment]
Thanks, but using the predict function, I get the contribution for each regressor for each date in the dataframe. I am trying to get the regression coefficients for the regressors I add to the model. In my case, I have added two regressors. However, the beta values in params have 28 values. I think 26 of these are for seasonality, and two of them are the regression coefficients. Is my understanding correct? However, which two of these correspond to regressors is something I am not able to figure out.

**Comment by bletham** [comment]
@AnupaS that is right, there are 26 coefficients for seasonality and 2 for the extra regressor. To see which coefficients correspond to which components, you can look at `m.train_component_cols`. This is a dataframe with 28 rows (one for each of the coefficients), and each column is a component (and is named for that component). The value in the dataframe is 1 if that beta is involved in that component, and 0 otherwise. For example, weekly seasonality uses 6 coefficients, and so you will see that the 'weekly' column has 6 1s. The extra regressors will use 1 each, and so you will find for their columns the one row that has a 1, and that is the index of the beta you want.

An important thing to note when looking at the regression coefficients is that Prophet does a lot of data normalization prior to model fitting. The `y` data that is giving in fit is scaled by its max value. Extra regressor columns are also by default standardized (that is, subtract mean, divide by sd) if non-binary. So the coefficient value that you see will be one that is applied to the standardized values of the extra regressors, and then added to the scaled trend forecast. The raw values of the coefficients thus may not be that useful to you.

**Comment by AnupaS** [comment]
@bletham Thanks for the help, that solves my issue.

**Comment by jarodleeyl** [comment]
Hi,

Just want to clarify and make sure I completely understand this. The code is written in R language.

I have included 6 regressors in the model. 

m = prophet() %>%
      add_regressor('no_hours_open') %>%
      add_regressor('easter_fri_indc') %>%
      add_regressor('easter_sat_indc') %>%
      add_regressor('easter_sun_indc') %>%
      add_regressor('easter_mon_indc') %>%
      add_regressor('anzac_day_indc') %>%
      fit.prophet(dat)

The model returns 12 coefficients.

> m$params$beta
           [,1]      [,2]       [,3]       [,4]        [,5]         [,6]       [,7]       [,8]       [,9]    [,10]     [,11]     [,12]
[1,] 0.03722567 0.0259044 0.01105772 -0.0102999 0.007605983 0.0007221441 0.09742383 -0.2358922 0.04465266 0.155911 0.1140141 0.1049005

And also > as.matrix(m$train.component.cols) returns a matrix with 12 rows and 10 columns.

To obtain the coefficients, I use the following:
> m$params$beta %*% as.matrix(m$train.component.cols)
     additive_terms anzac_day_indc easter_fri_indc easter_mon_indc easter_sat_indc easter_sun_indc extra_regressors_additive no_hours_open
[1,]       0.353226      0.1049005      -0.2358922       0.1140141      0.04465266        0.155911                 0.2810099    0.09742383
         weekly multiplicative_terms
[1,] 0.07221602                    0

Am I on the right track?

**Comment by bletham** [comment]
That looks right. There are 12 coefficients, 6 for the regressors you added and 6 for weekly seasonality.

**Comment by choudharyneelabh** [comment]
Hi, Im using R for Facebook Prophet and i keep getting the error message when i try to add a regressor:
"Error in add_regressor(model, name = "spend") : 
  Regressors must be added prior to model fitting."

Here's my code: I'm trying to add the column name "spend" as the regressor and the data frame i'm using to train the model by the way does indeed have spend it in it

full_2019_20 <- fread("raw_files/xx.csv") %>%
  select(-Day)%>%
  mutate(Date = mdy(Date)) %>%
  rename(ds = Date,
         y =`Overall Site Visit`,
         spend = `Media Cost`) %>%
  select(ds, y, spend)


model <- prophet(full_2019_20) 
model <- add_regressor(model, name = "spend")

**Comment by bletham** [comment]
You need to create the model without given it data so it doesn't fit:
```
model <- prophet()
model <- add_regressor(model, name="spend")
model <- fit.prophet(model, full_2019_20)
```
See https://facebook.github.io/prophet/docs/seasonality,_holiday_effects,_and_regressors.html#additional-regressors for the code example.

**Comment by carrasq** [comment]
Hi there-

A couple of questions on the transformations.

* Are the normalizations reversible? Or do we need to figure out the mean / sd of the regressors and the max of the `y` ahead of time? Right now I am using `StandardScaler` from `sklearn` and storing the key values for `y` and each regressor. I then reverse transform the output at the end. I'd rather have everything happening within Prophet. I also didn't know about the scaling of `y` at the time, so I'll need to figure that out.
* Out of curiosity, for the scaling of `y` - what is this for? Is it helping with outliers?

Appreciate the work - I'm having a lot of fun exploring the package.

**Comment by bletham** [comment]
Something to note: As of #1572, there is now in Python a utility for correctly extracting the regression coefficients of a model. This utility has shipped with v0.7.1 now in PyPI:
https://github.com/facebook/prophet/blob/77da5b8c06d05d478af510198c685d454342db1f/python/fbprophet/utilities.py#L28

It is used like:

```Python
from fbprophet.utilities import regressor_coefficients

regressor_df = regressor_coefficients(m)
```
It will properly account for all of the standardization happening under the hood.

@carrasq for your specific questions:
* Yes the normalizations are reversible. You can see the code in the utility above for how it is done, but basically the scaling on `y` is stored in `m.y_scale` and the standardization (mean and std) for each regressor is stored in the `m.extra_regressors` dict.
* Scaling y is to allow having reasonable priors in the model. The magnitudes of all of the priors (noise, trend changes, seasonality) are scale dependent, so normalizing the data makes it so default values can be reasonable across a larger set of time series.

**Comment by Mxsol** [comment]
@bletham Thank you for the update on the topic. Do I understand correctly that if I now need an incremental percentage change, such that 1% increase in X will result in beta% change in Y, I can just take the log of both Y and X  and the regressor_coefficients output will be the needed value?
As for now, my model coefficient without logs has negative sign, while a log-log attempt gives positive sign. I guess it might be due to noise in data. Can you suggest any other reasons apart from this, if there is only one regressor and multicollinearity is not the case?

**Comment by bletham** [comment]
@Mxsol  With an additive model,
```
yhat(t) = trend(t) + seasonality(t) + beta * regressor(t)
```
so if you use the utility to get the correctly transformed value of beta, then it means that a unit increase in `regressor` will produce a `beta` increase in `yhat`. With a multiplicative model,
```
yhat(t) = trend(t) * (1 + seasonality(t) + beta * regressor(t))
```
and a unit increase in `regressor` will produce a `beta * trend` increase in  `yhat`.

This is described also in the docstring for the utility:
https://github.com/facebook/prophet/blob/77da5b8c06d05d478af510198c685d454342db1f/python/fbprophet/utilities.py#L31-L35

If `beta` is negative, it means the model is fitting a negative impact of increasing `regressor` on `yhat`.

**Comment by AmbikaSowmyan** [comment]
> Something to note: As of #1572, there is now in Python a utility for correctly extracting the regression coefficients of a model. This utility has shipped with v0.7.1 now in PyPI:
> https://github.com/facebook/prophet/blob/77da5b8c06d05d478af510198c685d454342db1f/python/fbprophet/utilities.py#L28
> 
> It is used like:
> 
> ```python
> from fbprophet.utilities import regressor_coefficients
> 
> regressor_df = regressor_coefficients(m)
> ```
> 
> It will properly account for all of the standardization happening under the hood.
> 
> @carrasq for your specific questions:
> 
> * Yes the normalizations are reversible. You can see the code in the utility above for how it is done, but basically the scaling on `y` is stored in `m.y_scale` and the standardization (mean and std) for each regressor is stored in the `m.extra_regressors` dict.
> * Scaling y is to allow having reasonable priors in the model. The magnitudes of all of the priors (noise, trend changes, seasonality) are scale dependent, so normalizing the data makes it so default values can be reasonable across a larger set of time series.

Hey - when i try to  run this ->  from fbprophet.utilities import regressor_coefficients 
i get ModuleNotFoundError: No module named 'fbprophet.utilities'
any help on this is greatly appreciated

**Comment by bletham** [comment]
@AmbikaSowmyan can you check what version of fbprophet you're using? It will be there for 0.7 and later.
```
import fbprophet
print(fbprophet.__version__)
```
I'm guessing it's old, in which case you should upgrade (and note that for the latest version the package name has changed to `prophet`)

**Comment by AmbikaSowmyan** [comment]
> @AmbikaSowmyan can you check what version of fbprophet you're using? It will be there for 0.7 and later.
> 
> ```
> import fbprophet
> print(fbprophet.__version__)
> ```
> 
> I'm guessing it's old, in which case you should upgrade (and note that for the latest version the package name has changed to `prophet`)

@bletham yes thanks I had an older version of fbprophet which was the reason.

**Comment by imad24** [comment]
> @Mxsol With an additive model,
> 
> ```
> yhat(t) = trend(t) + seasonality(t) + beta * regressor(t)
> ```
> 
> so if you use the utility to get the correctly transformed value of beta, then it means that a unit increase in `regressor` will produce a `beta` increase in `yhat`. With a multiplicative model,
> 
> ```
> yhat(t) = trend(t) * (1 + seasonality(t) + beta * regressor(t))
> ```
> 
> and a unit increase in `regressor` will produce a `beta * trend` increase in `yhat`.
> 
> This is described also in the docstring for the utility:
> https://github.com/facebook/prophet/blob/77da5b8c06d05d478af510198c685d454342db1f/python/fbprophet/utilities.py#L31-L35
> 
> If `beta` is negative, it means the model is fitting a negative impact of increasing `regressor` on `yhat`.

Hey @bletham,
Sorry for opening this again but there is something I dont understand between the forecast dataframe returned by the predict function and the beta coefficients I get with the utility function.
Assuming a simple additive example model:
```python
#fbprophet.__version__ == 0.7.1
#Peyton Manning example
df = pd.read_csv(f'pmanning.csv', sep=';') 
#add an extra regressor with 3 values [1,2,3] for simplicity
df['extra_regressor']= pd.Series(np.random.choice([1,2,3], 2905), index=pd.to_datetime(df.ds))

#fit an additive prophet model
m = Prophet(daily_seasonality=False, interval_width=0, mcmc_samples=0, uncertainty_samples=0)
m.add_regressor('extra_regressor', mode = "additive")
f = m.fit(df).predict()
```
The forecast result looks like this;
![image](https://user-images.githubusercontent.com/1884358/121504323-37bf2e00-c9e2-11eb-917e-8427e8a3c91f.png)

While the regressor_coefficients are the following:
```python
from fbprophet.utilities import regressor_coefficients
beta = regressor_coefficients(m)
	regressor	regressor_mode	 center	       coef_lower	 coef	            coef_upper
0	extra_regressor	additive	1.993115	-0.012455	-0.012455	-0.012455
```
Knowing that the additive model equation is:
`yhat(t) = trend(t) + seasonality(t) + beta * regressor(t)`

And the column values in the dataframe do satisfy:
`f.yhat == (f.trend) + (f.weekly + f.yearly) + (f.extra_regressor) `

Shouldn't the extra_regressor column contain one of the 3 values (beta * regressor(t)) ?
```
-0.012455 * 1
-0.012455 * 2
-0.012455 * 3
```
My question is why are the values in the extra regressor column completely different from the beta coefficients ?
Where do the values in the dataframe (0.012369, -0.000086,-0.012541) come from ?

**Comment by KushalRunthala** [comment]
> > @Mxsol With an additive model,
> > ```
> > yhat(t) = trend(t) + seasonality(t) + beta * regressor(t)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so if you use the utility to get the correctly transformed value of beta, then it means that a unit increase in `regressor` will produce a `beta` increase in `yhat`. With a multiplicative model,
> > ```
> > yhat(t) = trend(t) * (1 + seasonality(t) + beta * regressor(t))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and a unit increase in `regressor` will produce a `beta * trend` increase in `yhat`.
> > This is described also in the docstring for the utility:
> > https://github.com/facebook/prophet/blob/77da5b8c06d05d478af510198c685d454342db1f/python/fbprophet/utilities.py#L31-L35
> > 
> > If `beta` is negative, it means the model is fitting a negative impact of increasing `regressor` on `yhat`.
> 
> Hey @bletham, Sorry for opening this again but there is something I dont understand between the forecast dataframe returned by the predict function and the beta coefficients I get with the utility function. Assuming a simple additive example model:
> 
> ```python
> #fbprophet.__version__ == 0.7.1
> #Peyton Manning example
> df = pd.read_csv(f'pmanning.csv', sep=';') 
> #add an extra regressor with 3 values [1,2,3] for simplicity
> df['extra_regressor']= pd.Series(np.random.choice([1,2,3], 2905), index=pd.to_datetime(df.ds))
> 
> #fit an additive prophet model
> m = Prophet(daily_seasonality=False, interval_width=0, mcmc_samples=0, uncertainty_samples=0)
> m.add_regressor('extra_regressor', mode = "additive")
> f = m.fit(df).predict()
> ```
> 
> The forecast result looks like this; ![image](https://user-images.githubusercontent.com/1884358/121504323-37bf2e00-c9e2-11eb-917e-8427e8a3c91f.png)
> 
> While the regressor_coefficients are the following:
> 
> ```python
> from fbprophet.utilities import regressor_coefficients
> beta = regressor_coefficients(m)
> 	regressor	regressor_mode	 center	       coef_lower	 coef	            coef_upper
> 0	extra_regressor	additive	1.993115	-0.012455	-0.012455	-0.012455
> ```
> 
> Knowing that the additive model equation is: `yhat(t) = trend(t) + seasonality(t) + beta * regressor(t)`
> 
> And the column values in the dataframe do satisfy: `f.yhat == (f.trend) + (f.weekly + f.yearly) + (f.extra_regressor) `
> 
> Shouldn't the extra_regressor column contain one of the 3 values (beta * regressor(t)) ?
> 
> ```
> -0.012455 * 1
> -0.012455 * 2
> -0.012455 * 3
> ```
> 
> My question is why are the values in the extra regressor column completely different from the beta coefficients ? Where do the values in the dataframe (0.012369, -0.000086,-0.012541) come from ?

@imad24 were you able to figure this out? I am facing similar issue
@bletham what is the explanation for this?

**Comment by elif-tr** [comment]
@imad24 @KushalRunthala @bletham 
I am actually having the same issue in this part of the output. I also tried to scale the value x[i] from the raw data with the mean and std I extracted for that variable from the output dictionary and multiplied with the coefficient produced by the model for that regressor but it still does not produce the value for that variable on that given day. Are you able to solve this issue?
